src/simsopt/geo/surfaceobjectives.py

```diff
@@ -141,6 +141,70 @@ def d2J_by_dsurfacecoefficientsdsurfacecoefficients(self):
         return out
 
 
+class PrincipalCurvature(Optimizable):
+    r"""
+
+    Given a Surface, evaluates a metric based on the principal curvatures,
+    :math:`\kappa_1` and :math:`\kappa_2`, where :math:`\kappa_1>\kappa_2`.
+    This metric is designed to penalize :math:`\kappa_1 > \kappa_{\max,1}` and
+    :math:`-\kappa_2 > \kappa_{\max,2}`.
+
+    .. math::
+       J &= \int d^2 x \exp \left(- ( \kappa_1 - \kappa_{\max,1})/w_1) \right) \\
+         &+ \int d^2 x \exp \left(- (-\kappa_2 - \kappa_{\max,2})/w_2) \right).
+
+    This metric can be used as a regularization within fixed-boundary optimization
+    to prevent, for example, surfaces with concave regions
+    (large values of :math:`|\kappa_2|`) or surfaces with large elongation
+    (large values of :math:`\kappa_1`).
+
+    """
+
+    def __init__(self, surface, kappamax1=1, kappamax2=1, weight1=0.05, weight2=0.05):
+        super().__init__(depends_on=[surface])
+        self.surface = surface
+        self.kappamax1 = kappamax1
+        self.kappamax2 = kappamax2
+        self.weight1 = weight1
+        self.weight2 = weight2
+        self.recompute_bell()
+
+    def J(self):
+        if (self._J is None):
+            self.compute()
+        return self._J
+
+    def dJ(self):
+        if (self._dJ is None):
+            self.compute()
+        return self._dJ
+
+    def recompute_bell(self, parent=None):
+        self._J = None
+        self._dJ = None
+
+    def compute(self):
```

Comment:
What is the reason for this structure? Isn't this very expensive if someone just wants to compute the objective?

Reply:
You are correct. I updated the structure.